### ISSUE TITLE
Duplicate clause in SQL query for files search

### ISSUE THREAD

**Issue by PVince81** [nextcloud/server#28564]
### Steps

1. Create two users "sender" and "receiver"
2. Login as "sender"
3. Create 5 folders "test1", "test2", ...
4. Share them all with "receiver"
5. Login as "receiver"
6. Search for "test" using the global search

### Expected result
Optimized query with only a single "storage" clause.
```
SELECT `file`.`fileid`, `storage`, `path`, `path_hash`, `file`.`parent`, `name`, `mimetype`, `mimepart`, `size`, `mtime`, `storage_mtime`, `encrypted`, `etag`, `permissions`, `checksum`, `metadata_etag`, `creation_time`, `upload_time` FROM `oc_filecache` `file` LEFT JOIN `oc_filecache_extended` `fe` ON `file`.`fileid` = `fe`.`fileid` WHERE (`name`  COLLATE utf8mb4_general_ci LIKE '%test%') AND (((`storage` = 4) AND ((`path_hash` = '45b963397aa40d4a0063e0d85e4fe7a1') OR (`path`  COLLATE utf8mb4_general_ci LIKE 'files/%'))) OR (`storage` = 3)) ORDER BY `mtime` desc LIMIT 5;
```

### Actual result
Query has 5 storage clauses with the same value.
The more shares are received, the more clauses there are.

```
SELECT `file`.`fileid`, `storage`, `path`, `path_hash`, `file`.`parent`, `name`, `mimetype`, `mimepart`, `size`, `mtime`, `storage_mtime`, `encrypted`, `etag`, `permissions`, `checksum`, `metadata_etag`, `creation_time`, `upload_time` FROM `oc_filecache` `file` LEFT JOIN `oc_filecache_extended` `fe` ON `file`.`fileid` = `fe`.`fileid` WHERE (`name`  COLLATE utf8mb4_general_ci LIKE '%test%') AND (((`storage` = 4) AND ((`path_hash` = '45b963397aa40d4a0063e0d85e4fe7a1') OR (`path`  COLLATE utf8mb4_general_ci LIKE 'files/%'))) OR (`storage` = 3) OR (`storage` = 3) OR (`storage` = 3) OR (`storage` = 3) OR (`storage` = 3)) ORDER BY `mtime` desc LIMIT 5;
```

### Version

Nextcloud 22.1.0

@icewind1991 as discussed, we should be able to deduplicate the `$caches` array here: https://github.com/nextcloud/server/blob/v22.1.0/lib/private/Files/Cache/QuerySearchHelper.php#L123

**Comment by icewind1991** [comment]
The `storage = 3` expressions are probably missing further filtering due to #28476

**Comment by PVince81** [comment]
furthermore I saw some repeated queries that we could probably batch too:
```
SELECT `path` FROM `*PREFIX*filecache` WHERE (`storage` = :dcValue1) AND (`fileid` = :dcValue2) with params dcValue1 => '3', dcValue2 => '82'
SELECT `path` FROM `*PREFIX*filecache` WHERE (`storage` = :dcValue1) AND (`fileid` = :dcValue2) with params dcValue1 => '3', dcValue2 => '83'
SELECT `path` FROM `*PREFIX*filecache` WHERE (`storage` = :dcValue1) AND (`fileid` = :dcValue2) with params dcValue1 => '3', dcValue2 => '84'
SELECT `path` FROM `*PREFIX*filecache` WHERE (`storage` = :dcValue1) AND (`fileid` = :dcValue2) with params dcValue1 => '3', dcValue2 => '85'
SELECT `path` FROM `*PREFIX*filecache` WHERE (`storage` = :dcValue1) AND (`fileid` = :dcValue2) with params dcValue1 => '3', dcValue2 => '86'
SELECT `path` FROM `*PREFIX*filecache` WHERE (`storage` = :dcValue1) AND (`fileid` = :dcValue2) with params dcValue1 => '3', dcValue2 => '87'
```
at least batch when on the same storage

**Comment by PVince81** [comment]
> The `storage = 3` expressions are probably missing further filtering due to #28476

that PR was merged so not sure if related, or maybe you simply wanted to point at the post-filtering code?

**Comment by slater0013** [comment]
NC 21.0.5 here !

Search function seems to be written completely differently but the problem is the same !

One more detail : 

step 3 / create **file** and not **folder** testfile1
step 4/ share with receiver 
step 6/ search testfile

In slow query log, I can see twice : 
`
SELECT ***fields*** 
FROM ***prefix_filecache***  LEFT JOIN ***prefix_filecache_extended*** ON ***fileid*** 
WHERE (storage = ***sender_storage_id***) 
AND (
 (
  (path  COLLATE utf8mb4_general_ci LIKE '***complete_path_of_the_shared_file***/%') 
  OR 
  (path_hash = ***path_hash***)
 ) 
 AND (name  COLLATE utf8mb4_general_ci LIKE '%***my_search_pattern***%')
) 
ORDER BY mtime desc LIMIT 5;
`

The system is trying to search my pattern under the "complete/path/to/the/shared/file" adding a trailing slash like if It was a folder !

In oc_share, this share has item_type = "file" so the system is able to know that this is a file and not a folder. It should not loose time searching in impossible paths.

Hope It's clear...

GS

**Comment by szaimen** [comment]
Hi, please update to 24.0.9 or better 25.0.3 and report back if it fixes the issue. Thank you!

My goal is to add a label like e.g. 25-feedback to this ticket of an up-to-date major Nextcloud version where the bug could be reproduced. However this is not going to work without your help. So thanks for all your effort!

If you don't manage to reproduce the issue in time and the issue gets closed but you can reproduce the issue afterwards, feel free to create a new bug report with up-to-date information by following this link: https://github.com/nextcloud/server/issues/new?assignees=&labels=bug%2C0.+Needs+triage&template=BUG_REPORT.yml&title=%5BBug%5D%3A+

**Comment by slater0013** [comment]
was fixed before 24.0.9

**Comment by szaimen** [comment]
Thanks for verifying!
